orttraining/orttraining/python/orttraining_python_module.cc

```diff
@@ -183,35 +178,39 @@ void ORTTrainingPythonEnv::ClearExecutionProviderInstances() {
   execution_provider_instances_map_.clear();
 }
 
-static std::unique_ptr<ORTTrainingPythonEnv> ort_training_env;
-
-void InitializeTrainingEnv() {
-  auto initialize = [&]() {
-    static bool initialized = false;
-    if (initialized) {
-      return;
-    }
-    // Initialization of the module
+class TrainingEnvInitialzer {
+ public:
+  TrainingEnvInitialzer() {
     InitArray();
     Env::Default().GetTelemetryProvider().SetLanguageProjection(OrtLanguageProjection::ORT_PROJECTION_PYTHON);
-    ort_training_env = std::make_unique<ORTTrainingPythonEnv>();
-    initialized = true;
-  };
-  initialize();
-}
+    ort_training_env_ = std::make_unique<ORTTrainingPythonEnv>();
+  }
 
-ORTTrainingPythonEnv& GetTrainingEnv() {
-  if (!ort_training_env) {
-    InitializeTrainingEnv();
+  ~TrainingEnvInitialzer() {
+    destroyed = true;
   }
-  return *ort_training_env;
-}
 
-Environment& GetTrainingORTEnv() {
-  if (!ort_training_env) {
-    InitializeTrainingEnv();
+  ORTTrainingPythonEnv& Get() noexcept {
+    return *ort_training_env_;
   }
-  return ort_training_env->GetORTEnv();
+
+  static bool destroyed;
+
+ private:
+  std::unique_ptr<ORTTrainingPythonEnv> ort_training_env_;
+};
+
+bool TrainingEnvInitialzer::destroyed = false;
+
+ORTTrainingPythonEnv& GetTrainingEnv() {
+  // Guard against attempts to resurrect the singleton
+  if (TrainingEnvInitialzer::destroyed) {
+    ORT_THROW("Detected an attempt to resurrect destroyed Training Env Environment");
+  }
```

Comment:
Can this ever actually happen?

Reply:
Yes. For a detailed discussion on global objects and singletons see `Modern C++ Design` by A. Alexandrescu.